### PR TITLE
Indexing within indicators

### LINKED ISSUE #899: Masking and indexing as argument of indicator's call
Proposition to inject the `**indexer` args to _all_ indicators. 

With this proposition, `[Resampling]Indicator`'s call would mask values along the `time` axis according to the instruction passed to the `indexer` kwargs. The masking would occur before the compute. The `missing` methods would need to be modified to take this masking into account.

If possible, the masked values could even be dropped, in order to accelerate computation.

For example:  `xc.atmos.tg_max(tas, season='MAM', freq='YS')` would return one value per year (on Jan 1st) of the maximal temperature in the spring months. Internally, it would look like `tas2 = tas.where(tas.time.dt.month.isin([3, 4, 5]))`.

But we could open up to fancier indexing: 
Ex: `xc.atmos.tg_max(tas, within_dates=['03-13', '05-14'], freq='YS')`. Again yearly values of maximal temperature but only considering days between March 13th and May 14th. 

Maybe even : 
```
start = xc.atmos.first_day_above(...)
end = xc.atmos.first_snowfall(...)
xc.atmos.tg_max(tas, within)_dates=(start, end), freq='YS')
```
?

I think this needs:

- [ ] A standalone function to mask a timeseries according to a tuple of dates (doy int , month-day str, arrays of doys). @yrouranos has a draft of this in #894. @RondeauG  has also a need for this. 
- [ ]  #867 to be done.
- [ ] A way for `missing` methods to take this into account.

It would fix a reccuring issues of users wanting to compute some indicators on clea-but-incomplete data or on subsets. It would on other hand either fill the signature with many arguments or, if we leave them as kwargs, forbid the use of kwargs by indices (which is not too bad).

### PR DESCRIPTION
**Pull request by aulemahal**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #899
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added.
- [x] `bumpversion patch` has been called on this branch
- [x] The relevant author information has been added to `.zenodo.json`

### What kind of change does this PR introduce?

* Creates another base Indicator class, `ResamplingIndicatorWithIndexing`, that injects the `indexer` kwargs, as it was done in some "stats" function. The indexing is done in the `preprocess_and_check` function, so the indice (`compute`) receives an input where values outside the select period are **masked**.
* Expand `generic.select_time` with `doy_bounds` and `date_bounds`. The first takes in bounds as day-of-year integers and the second takes in dates in the month-day format (`%m-%d`). 

### Does this PR introduce a breaking change?

* Indexing now _masks_ values by default, instead of dropping them. There are issues when values are dropped, some indices do not expect those gaps. The principal culprit is `units.rate2amount` which converts by multiplying the rate by the timestep length, which is inferred from the coordinate itself. In case of a gap, the rate is assumed constant on the whole gap...
* Most indicators now have the new `**indexer`  in their signature.

### Other information:
An indice may add an `**indexer` var_keyword and handle the time selecting itself. If the indicator is constructed from `ResamplingIndicator` (the one that _doesn't_ inject this new param), the missing values handling will still understand the passed indexing parameter. This can be useful when the order between indexing and computing is important (like for #884 and many others).
This means that the order is currently "wrong" for many cases: i.e. indexing is done before computation. In a way, if it is made clear what the default order is, I think it's ok to have this caveat temporarily. But should we: 
1. Implement these cases now, in this PR.
2. Let them be wrong for now and correct them later.
3. Revert injecting `**indexer` everywhere and slowly implement it where it makes most sense afterwards.
?

### CODE REVIEW

File: xclim/core/indicator.py [hunk 1]
```diff
@@ -1309,30 +1331,64 @@ def _postprocess(self, outs, das, params, indexer):
...
+                self._missing(
+                    da, params["freq"], src_freq, options, params.get("indexer", {})
+                )
                 for da in das.values()
                 if "time" in da.coords
             )
-            mask = reduce(np.logical_or, miss)
+            # Reduce by or and broadcast to ensure the same length in time
+            # When indexing is used and there are no valid points in the last period, mask will not include it
+            mask = reduce(np.logical_or, miss).reindex_like(outs[0], fill_value=True)
```
Comment (aulemahal) [inline xclim/core/indicator.py:1342]: In the case where the last period contains no points after indexing, it doesn't appear in `mask` and so it was dropped from the result, which is a breaking change in a way. 
However, this `reindex_like` trick adds complexity, but fixes the problem.

I would prefer a breaking change than this.

Ex: If you do something with a time series from Jan 1st to Dec 31st, then pass `freq='AS=JUL'` and indexer : `month=2`. The second period, is completely dropped from the mask as it contains no points. Before, it wasn't dropped, rather it was masked out.

File: xclim/indices/generic.py [hunk 2]
```diff
@@ -57,31 +59,120 @@
...
+
+        # Get doy of date, this is now safe because the calendar is uniform.
+        start = to_cftime_datetime("2000-" + start, calendar).dayofyr
+        end = to_cftime_datetime("2000-" + end, calendar).dayofyr
+
+        if start <= end:
+            doys = np.arange(start, end + 1)
+        else:
+            doys = np.concatenate((np.arange(start, 367), np.arange(0, end + 1)))
+        mask = time.time.dt.dayofyear.isin(doys)
```
Comment (huard) [inline xclim/indices/generic.py:172]: Could you just set `doy_bounds` and move the `elif doy_bounds is not None:` code segment below to avoid repeating ?

**Review by huard — APPROVED** [review]

File: xclim/core/indicator.py [hunk 3]
```diff
@@ -1133,13 +1155,13 @@ def _format(
...
-            mba["indexer"] = dv
-        else:
-            mba["indexer"] = "annual"
+        # if indexer:
+        #     dk, dv = indexer.copy().popitem()
+        #     if dk == "month":
+        #         dv = "m{}".format(dv)
+        #     mba["indexer"] = dv
+        # else:
+        #     mba["indexer"] = "annual"
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xclim/core/indicator.py:1164]: Code is semi obsolete, but functionality needs to be implemented back (that's the failing tests).

[the remark this answers is not in the document]
Comment (aulemahal) [reply xclim/core/indicator.py]: Updated, please have a look.

I leave to the AttrFormatter the role of translating the "m{X}", season or freq string into words. 
Current code has the issue that `doy_bounds` and  `date_bounds` are suboptimally stated and currently not translatable.

**Review by huard — APPROVED** [review]

File: xclim/core/indicator.py [hunk 4]
```diff
@@ -1131,15 +1153,18 @@ def _format(
...
+            elif k == "indexer":
+                if v and v not in [_empty, _empty_default]:
+                    dk, dv = v.copy().popitem()
+                    if dk == "month":
+                        dv = f"m{dv}"
+                    elif dk in ("doy_bounds", "date_bounds"):
+                        dv = f"{dv[0]} to {dv[1]}"
+                    mba["indexer"] = dv
+                else:
+                    mba["indexer"] = args.get("freq") or indices.generic.default_freq()
```
Comment (aulemahal) [inline xclim/core/indicator.py:1165]: Woups, I realize the mistake. `v` is not a correct indexer, that's the point!

But then, I'm not sure of the need of `default_freq`? 
```suggestion
                    mba["indexer"] = args.get("freq") or "YS"
```

**Pull request merged**